README.md
<div align="center">
  <h1>types-pkg-json</h1>
  <a href="https://npmjs.com/package/types-pkg-json">
    <img alt="NPM" src="https://img.shields.io/npm/v/types-pkg-json.svg">
  </a>
  <a href="https://github.com/bconnorwhite/types-pkg-json">
    <img alt="TypeScript" src="https://img.shields.io/github/languages/top/bconnorwhite/types-pkg-json.svg">
  </a>
  <a href='https://coveralls.io/github/bconnorwhite/types-pkg-json?branch=master'>
    <img alt="Coverage Status" src="https://img.shields.io/coveralls/github/bconnorwhite/types-pkg-json.svg?branch=master">
  </a>
  <a href="https://github.com/bconnorwhite/types-pkg-json">
    <img alt="GitHub Stars" src="https://img.shields.io/github/stars/bconnorwhite/types-pkg-json?label=Stars%20Appreciated%21&style=social">
  </a>
  <a href="https://twitter.com/bconnorwhite">
    <img alt="Twitter Follow" src="https://img.shields.io/twitter/follow/bconnorwhite.svg?label=%40bconnorwhite&style=social">
  </a>
</div>

<br />

> Type checking for package.json.

Derived from [type-fest](https://www.npmjs.com/package/type-fest) with a focus on package.json types.

## Installation

```bash
yarn add types-pkg-json
```

```bash
npm install types-pkg-json
```

## API

### Types

```ts
import { fileName, PackageJSON } from "types-pkg-json"; // Type for NPM's package.json file

console.log(fileName); // "package.json"
```

#### Field Types

```ts
import {
  BugsLocation, // Used by the `bugs` field.
  LicenseID, // Used by the `license` and `licenses.type` fields.
  Person, // Used by the `author`, `contributors`, and `maintainers` fields.
  DirectoryLocations, // Used by the `directories` field.
  Repository, // Used by the `repository` field.
  Scripts, // Used by the `scripts` field.
  Dependencies // Used by the `dependencies`, `devDependencies`, `optionalDependencies`, and `peerDependencies` fields.
} from "types-pkg-json";
```

#### Configuration Types

```ts
import {
  NonStandardEntryPoints, // module, esnext, browser, sideEffects
  TypeScriptConfiguration, // types, typings
  WorkspaceConfig, // packages, nohoist
  YarnConfiguration, // workspaces, flat, resolutions
  JSPMConfiguration, // jspm
  ESLintConfiguration // eslintConfig
} from "types-pkg-json";
```

#### Utility Types

```ts
import {
  isJSONObject
  JSONObject,
  JSONValue,
  JSONArray,
  Primitive,
  LiteralUnion
} from "types-pkg-json";

function isJSONObject(object?: JSONValue): object is JSONObject;
```

<br />

<h2>Dependencies<img align="right" alt="dependencies" src="https://img.shields.io/david/bconnorwhite/types-pkg-json.svg"></h2>

- [types-json](https://www.npmjs.com/package/types-json): Type checking for JSON objects
- [types-eslintrc](https://www.npmjs.com/package/types-eslintrc): Type checking for .eslintrc.json

<br />

<h2>Dev Dependencies<img align="right" alt="David" src="https://img.shields.io/david/dev/bconnorwhite/types-pkg-json.svg"></h2>

- [@bconnorwhite/bob](https://www.npmjs.com/package/@bconnorwhite/bob): Bob is a toolkit for TypeScript projects

<h2>License <img align="right" alt="license" src="https://img.shields.io/npm/l/types-pkg-json.svg"></h2>

[MIT](https://opensource.org/licenses/MIT)

<br />

## Related Projects

- [types-json](https://www.npmjs.com/package/types-json): Type checking for JSON objects
- [types-tsconfig](https://www.npmjs.com/package/types-tsconfig): Type checking for tsconfig.json
- [types-eslintrc](https://www.npmjs.com/package/types-eslintrc): Type checking for .eslintrc.json


CHANGELOG.md
# [1.2.0](https://github.com/bconnorwhite/types-package-json/compare/v1.1.4...v1.2.0) (2020-10-09)


### Features

* export PackageJSONFileName type ([2dedb04](https://github.com/bconnorwhite/types-package-json/commit/2dedb04c11efc80e884bc1b04a6bd959f52665d4))



## [1.1.4](https://github.com/bconnorwhite/types-package-json/compare/v1.1.3...v1.1.4) (2020-10-08)



## [1.1.3](https://github.com/bconnorwhite/types-package-json/compare/v1.1.2...v1.1.3) (2020-10-08)



## [1.1.2](https://github.com/bconnorwhite/types-package-json/compare/v1.1.1...v1.1.2) (2020-09-19)



## [1.1.1](https://github.com/bconnorwhite/types-package-json/compare/v1.1.0...v1.1.1) (2020-09-19)



# [1.1.0](https://github.com/bconnorwhite/types-package-json/compare/v1.0.2...v1.1.0) (2020-09-09)



## [1.0.2](https://github.com/bconnorwhite/types-package-json/compare/v1.0.1...v1.0.2) (2020-09-08)



## 1.0.1 (2020-09-08)





test/index.test.ts
import { test, expect } from "@jest/globals";
import { fileName, isJSONObject } from "../source";

test("import sanity check", () => {
  expect(fileName).toBe("package.json");
});

test("isJSONObject", () => {
  expect(isJSONObject({ name: "package name", description: "description", version: "1.0.0" })).toBe(true);
});


source/utils.ts

/**
 * Matches any [primitive value](https://developer.mozilla.org/en-US/docs/Glossary/Primitive).
 */
export type Primitive =
  | null
  | undefined
  | string
  | number
  | boolean
  | symbol
  | bigint;

/**
 * Allows creating a union type by combining primitive types and literal types without sacrificing auto-completion in IDEs for the literal type part of the union.
 * Currently, when a union type of a primitive type is combined with literal types, TypeScript loses all information about the combined literals.
 * Thus, when such type is used in an IDE with autocompletion, no suggestions are made for the declared literals.
 * This type is a workaround for [Microsoft/TypeScript#29729](https://github.com/Microsoft/TypeScript/issues/29729). It will be removed as soon as it's not needed anymore.
 */
export type LiteralUnion<LiteralType extends BaseType, BaseType extends Primitive> =
  | LiteralType
  | (BaseType & {_?: never});


source/package-json.ts
import { JSONObject } from "types-json";
import { LiteralUnion } from "./utils";
import {
  BugsLocation,
  LicenseID,
  Licenses,
  Person,
  Bin,
  DirectoryLocations,
  Repository,
  Scripts,
  Dependencies,
  PeerDependenciesMeta,
  Engines,
  OS,
  CPU,
  Funding
} from "./fields";
import {
  YarnConfiguration,
  TypeScriptConfiguration,
  JSPMConfiguration,
  ESLintConfiguration
} from "./configuration";

export type NonStandardEntryPoints = {
  /**
   * An ECMAScript module ID that is the primary entry point to the program.
   */
  module?: string;
  /**
   * A module ID with untranspiled code that is the primary entry point to the program.
   */
  esnext?: string | {
    [moduleName: string]: string | undefined;
    main?: string;
    browser?: string;
  };
  /**
   * A hint to JavaScript bundlers or component tools when packaging modules for client side use.
   */
  browser?: string | {
    [moduleName: string]: string | false;
  };
  /**
   * Denote which files in your project are "pure" and therefore safe for Webpack to prune if unused.
   * [Read more.](https://webpack.js.org/guides/tree-shaking/)
   */
  sideEffects?: boolean | string[];
}

/**
 * Type for [npm's `package.json` file](https://docs.npmjs.com/creating-a-package-json-file).
 * Also includes types for fields used by other popular projects, like TypeScript and Yarn.
 */
export type PackageJSON = JSONObject & {
  /**
   * The name of the package.
   */
  name?: string;
  /**
   * Package version, parseable by [`node-semver`](https://github.com/npm/node-semver).
   */
  version?: string;
  /**
   * Package description, listed in `npm search`.
   */
  description?: string;
  /**
   * Keywords associated with package, listed in `npm search`.
   */
  keywords?: string[];
  /**
   * The URL to the package's homepage.
   */
  homepage?: LiteralUnion<".", string>;
  /**
   * The URL to the package's issue tracker and/or the email address to which issues should be reported.
   */
  bugs?: BugsLocation;
  /**
   * The license for the package.
   */
  license?: LicenseID;
  /**
   * The licenses for the package.
   */
  licenses?: Licenses;
  author?: Person;
  /**
   * A list of people who contributed to the package.
   */
  contributors?: Person[];
  /**
   * A list of people who maintain the package.
   */
  maintainers?: Person[];
  /**
   * The files included in the package.
   */
  files?: string[];
  /**
   * The module ID that is the primary entry point to the program.
   */
  main?: string;
  /**
   * The executable files that should be installed into the `PATH`.
   */
  bin?: Bin;
  /**
   * Filenames to put in place for the `man` program to find.
   */
  man?: string | string[];
  /**
   * Indicates the structure of the package.
   */
  directories?: DirectoryLocations;
  /**
   * Location for the code repository.
   */
  repository?: Repository;
  /**
   * Script commands that are run at various times in the lifecycle of the package. The key is the lifecycle event, and the value is the command to run at that point.
   */
  scripts?: Scripts;
  /**
   * Is used to set configuration parameters used in package scripts that persist across upgrades.
   */
  config?: JSONObject;
  /**
   * The dependencies of the package.
   */
  dependencies?: Dependencies;
  /**
   * Additional tooling dependencies that are not required for the package to work. Usually test, build, or documentation tooling.
   */
  devDependencies?: Dependencies;
  /**
   * Dependencies that are skipped if they fail to install.
   */
  optionalDependencies?: Dependencies;
  /**
   * Dependencies that will usually be required by the package user directly or via another dependency.
   */
  peerDependencies?: Dependencies;
  /**
   * Indicate peer dependencies that are optional.
   */
  peerDependenciesMeta?: PeerDependenciesMeta;
  /**
   * Package names that are bundled when the package is published.
   */
  bundledDependencies?: string[];
  /**
   * Alias of `bundledDependencies`.
   */
  bundleDependencies?: string[];
  /**
   * Engines that this package runs on.
   */
  engines?: Engines;
  /**
   * @deprecated
   */
  engineStrict?: boolean;
  /**
   * Operating systems the module runs on.
   */
  os?: OS[];
  /**
   * CPU architectures the module runs on.
   */
  cpu?: CPU[];
  /**
   * If set to `true`, a warning will be shown if package is installed locally. Useful if the package is primarily a command-line application that should be installed globally.
   * @deprecated
   */
  preferGlobal?: boolean;
  /**
   * If set to `true`, then npm will refuse to publish it.
   */
  private?: boolean;
  /**
   * A set of config values that will be used at publish-time. It's especially handy to set the tag, registry or access, to ensure that a given package is not tagged with 'latest',
   * published to the global public registry or that a scoped module is private by default.
   */
  publishConfig?: JSONObject;
  /**
   * Describes and notifies consumers of a package's monetary support information.
  [Read more.](https://github.com/npm/rfcs/blob/latest/accepted/0017-add-funding-support.md)
   */
  funding?: Funding;
} &
NonStandardEntryPoints &
TypeScriptConfiguration &
ESLintConfiguration &
YarnConfiguration &
JSPMConfiguration &
JSONObject;


source/fields/scripts.ts

export type Scripts = {
  /**
   * Run **before** the package is published (Also run on local `npm install` without any arguments).
   */
  prepublish?: string;
  /**
   * Run both **before** the package is packed and published, and on local `npm install` without any arguments. This is run **after** `prepublish`, but **before** `prepublishOnly`.
   */
  prepare?: string;
  /**
   * Run **before** the package is prepared and packed, **only** on `npm publish`.
   */
  prepublishOnly?: string;
  /**
   * Run **before** a tarball is packed (on `npm pack`, `npm publish`, and when installing git dependencies).
   */
  prepack?: string;
  /**
   * Run **after** the tarball has been generated and moved to its final destination.
   */
  postpack?: string;
  /**
   * Run **after** the package is published.
   */
  publish?: string;
  /**
   * Run **after** the package is published.
   */
  postpublish?: string;
  /**
   * Run **before** the package is installed.
   */
  preinstall?: string;
  /**
   * Run **after** the package is installed.
   */
  install?: string;
  /**
   * Run **after** the package is installed and after `install`.
   */
  postinstall?: string;
  /**
   * Run **before** the package is uninstalled and before `uninstall`.
   */
  preuninstall?: string;
  /**
   * Run **before** the package is uninstalled.
   */
  uninstall?: string;
  /**
   * Run **after** the package is uninstalled.
   */
  postuninstall?: string;
  /**
   * Run **before** bump the package version and before `version`.
   */
  preversion?: string;
  /**
   * Run **before** bump the package version.
   */
  version?: string;
  /**
   * Run **after** bump the package version.
   */
  postversion?: string;
  /**
   * Run with the `npm test` command, before `test`.
   */
  pretest?: string;
  /**
   * Run with the `npm test` command.
   */
  test?: string;
  /**
   * Run with the `npm test` command, after `test`.
   */
  posttest?: string;
  /**
   * Run with the `npm stop` command, before `stop`.
   */
  prestop?: string;
  /**
   * Run with the `npm stop` command.
   */
  stop?: string;
  /**
   * Run with the `npm stop` command, after `stop`.
   */
  poststop?: string;
  /**
   * Run with the `npm start` command, before `start`.
   */
  prestart?: string;
  /**
   * Run with the `npm start` command.
   */
  start?: string;
  /**
   * Run with the `npm start` command, after `start`.
   */
  poststart?: string;
  /**
   * Run with the `npm restart` command, before `restart`. Note: `npm restart` will run the `stop` and `start` scripts if no `restart` script is provided.
   */
  prerestart?: string;
  /**
   * Run with the `npm restart` command. Note: `npm restart` will run the `stop` and `start` scripts if no `restart` script is provided.
   */
  restart?: string;
  /**
   * Run with the `npm restart` command, after `restart`. Note: `npm restart` will run the `stop` and `start` scripts if no `restart` script is provided.
   */
  postrestart?: string;
} & {
  [scriptName in string]?: string;
};


source/configuration/yarn.ts
import { Dependencies } from "../fields";

/**
 * An alternative configuration for Yarn workspaces.
 */
export type WorkspaceConfig = {
  /**
   * An array of workspace pattern strings which contain the workspace packages.
   */
  packages?: WorkspacePattern[];
  /**
   * Designed to solve the problem of packages which break when their `node_modules` are moved to the root workspace directory - a process known as hoisting.
   * For these packages, both within your workspace, and also some that have been installed via `node_modules`,
   * it is important to have a mechanism for preventing the default Yarn workspace behavior.
   * By adding workspace pattern strings here, Yarn will resume non-workspace behavior for any package which matches the defined patterns.
   * [Read more](https://classic.yarnpkg.com/blog/2018/02/15/nohoist/)
   */
  nohoist?: WorkspacePattern[];
}

/**
 * A workspace pattern points to a directory or group of directories which contain packages that should be included in the workspace installation process.
 * The patterns are handled with [minimatch](https://github.com/isaacs/minimatch).
 * @example
 * `docs` → Include the docs directory and install its dependencies.
 * `packages/*` → Include all nested directories within the packages directory, like `packages/cli` and `packages/core`.
 */
export type WorkspacePattern = string;

export type YarnConfiguration = {
  /**
   * Used to configure [Yarn workspaces](https://classic.yarnpkg.com/docs/workspaces/).
   * Workspaces allow you to manage multiple packages within the same repository in such a way that you only need to run `yarn install`
   * once to install all of them in a single pass. Please note that the top-level `private` property of `package.json` **must** be set to `true` in order to use workspaces.
   */
  workspaces?: WorkspacePattern[] | WorkspaceConfig;
  /**
   * If your package only allows one version of a given dependency, and you’d like to enforce the same behavior as `yarn install --flat` on the command-line, set this to `true`.
   * Note that if your `package.json` contains `"flat": true` and other packages depend on yours (e.g. you are building a library rather than an app),
   * those other packages will also need `"flat": true` in their `package.json` or be installed with `yarn install --flat` on the command-line.
   */
  flat?: boolean;
  /**
   * Selective version resolutions. Allows the definition of custom package versions inside dependencies without manual edits in the `yarn.lock` file.
   */
  resolutions?: Dependencies;
}


source/types.ts
import { JSONObject } from "types-json";
import { ESLintConfig } from "types-eslintrc";

export {
  JSONObject
}

/**
Matches any [primitive value](https://developer.mozilla.org/en-US/docs/Glossary/Primitive).
*/
export type Primitive =
	| null
	| undefined
	| string
	| number
	| boolean
	| symbol
  | bigint;
  
/**
Allows creating a union type by combining primitive types and literal types without sacrificing auto-completion in IDEs for the literal type part of the union.
Currently, when a union type of a primitive type is combined with literal types, TypeScript loses all information about the combined literals. Thus, when such type is used in an IDE with autocompletion, no suggestions are made for the declared literals.
This type is a workaround for [Microsoft/TypeScript#29729](https://github.com/Microsoft/TypeScript/issues/29729). It will be removed as soon as it's not needed anymore.
 */
export type LiteralUnion<LiteralType extends BaseType, BaseType extends Primitive> = LiteralType | (BaseType & {_?: never});

/**
A person who has been involved in creating or maintaining the package.
*/
export type Person = string | {
	name: string;
	url?: string;
	email?: string;
};

export type BugsLocation = string | {
	/**
	The URL to the package's issue tracker.
	*/
	url?: string;

	/**
	The email address to which issues should be reported.
	*/
	email?: string;
};

export type DirectoryLocations = {
  /**
  Location for executable scripts. Sugar to generate entries in the `bin` property by walking the folder.
  */
  bin?: string;

  /**
  Location for Markdown files.
  */
  doc?: string;

  /**
  Location for example scripts.
  */
  example?: string;

  /**
  Location for the bulk of the library.
  */
  lib?: string;

  /**
  Location for man pages. Sugar to generate a `man` array by walking the folder.
  */
  man?: string;

  /**
  Location for test files.
  */
  test?: string;
}

export type Scripts = {
  /**
  Run **before** the package is published (Also run on local `npm install` without any arguments).
  */
  prepublish?: string;

  /**
  Run both **before** the package is packed and published, and on local `npm install` without any arguments. This is run **after** `prepublish`, but **before** `prepublishOnly`.
  */
  prepare?: string;

  /**
  Run **before** the package is prepared and packed, **only** on `npm publish`.
  */
  prepublishOnly?: string;

  /**
  Run **before** a tarball is packed (on `npm pack`, `npm publish`, and when installing git dependencies).
  */
  prepack?: string;

  /**
  Run **after** the tarball has been generated and moved to its final destination.
  */
  postpack?: string;

  /**
  Run **after** the package is published.
  */
  publish?: string;

  /**
  Run **after** the package is published.
  */
  postpublish?: string;

  /**
  Run **before** the package is installed.
  */
  preinstall?: string;

  /**
  Run **after** the package is installed.
  */
  install?: string;

  /**
  Run **after** the package is installed and after `install`.
  */
  postinstall?: string;

  /**
  Run **before** the package is uninstalled and before `uninstall`.
  */
  preuninstall?: string;

  /**
  Run **before** the package is uninstalled.
  */
  uninstall?: string;

  /**
  Run **after** the package is uninstalled.
  */
  postuninstall?: string;

  /**
  Run **before** bump the package version and before `version`.
  */
  preversion?: string;

  /**
  Run **before** bump the package version.
  */
  version?: string;

  /**
  Run **after** bump the package version.
  */
  postversion?: string;

  /**
  Run with the `npm test` command, before `test`.
  */
  pretest?: string;

  /**
  Run with the `npm test` command.
  */
  test?: string;

  /**
  Run with the `npm test` command, after `test`.
  */
  posttest?: string;

  /**
  Run with the `npm stop` command, before `stop`.
  */
  prestop?: string;

  /**
  Run with the `npm stop` command.
  */
  stop?: string;

  /**
  Run with the `npm stop` command, after `stop`.
  */
  poststop?: string;

  /**
  Run with the `npm start` command, before `start`.
  */
  prestart?: string;

  /**
  Run with the `npm start` command.
  */
  start?: string;

  /**
  Run with the `npm start` command, after `start`.
  */
  poststart?: string;

  /**
  Run with the `npm restart` command, before `restart`. Note: `npm restart` will run the `stop` and `start` scripts if no `restart` script is provided.
  */
  prerestart?: string;

  /**
  Run with the `npm restart` command. Note: `npm restart` will run the `stop` and `start` scripts if no `restart` script is provided.
  */
  restart?: string;

  /**
  Run with the `npm restart` command, after `restart`. Note: `npm restart` will run the `stop` and `start` scripts if no `restart` script is provided.
  */
  postrestart?: string;
} & {
  [scriptName in string]?: string;
};

/**
Dependencies of the package. The version range is a string which has one or more space-separated descriptors. Dependencies can also be identified with a tarball or Git URL.
*/
export type Dependencies = {
  [packageName: string]: string;
}

/**
Specify the place where your code lives. This is helpful for people who want to contribute.
*/
export type Repository = string | {
  type: LiteralUnion<"git" | "svn", string>;
  /**
  A publicly available (perhaps read-only) url that can be handed directly to a VCS program without any modification.
  */
  url: string;
  /**
  Relative path to package.json if it is placed in non-root directory (for example if it is part of a monorepo).
  [Read more.](https://github.com/npm/rfcs/blob/latest/implemented/0010-monorepo-subdirectory-declaration.md)
  */
  directory?: string;
};

export type NonStandardEntryPoints = {
  /**
  An ECMAScript module ID that is the primary entry point to the program.
  */
  module?: string;

  /**
  A module ID with untranspiled code that is the primary entry point to the program.
  */
  esnext?: string | {
    [moduleName: string]: string | undefined;
    main?: string;
    browser?: string;
  };

  /**
  A hint to JavaScript bundlers or component tools when packaging modules for client side use.
  */
  browser?: string | {
    [moduleName: string]: string | false;
  };

  /**
  Denote which files in your project are "pure" and therefore safe for Webpack to prune if unused.
  [Read more.](https://webpack.js.org/guides/tree-shaking/)
  */
  sideEffects?: boolean | string[];
}

export {
	ESLintConfig
}

export type TypeScriptConfiguration = {
  /**
  Location of the bundled TypeScript declaration file.
  */
  types?: string;

  /**
  Location of the bundled TypeScript declaration file. Alias of `types`.
  */
  typings?: string;
}

/**
An alternative configuration for Yarn workspaces.
*/
export type WorkspaceConfig = {
  /**
  An array of workspace pattern strings which contain the workspace packages.
  */
  packages?: WorkspacePattern[];

  /**
  Designed to solve the problem of packages which break when their `node_modules` are moved to the root workspace directory - a process known as hoisting. For these packages, both within your workspace, and also some that have been installed via `node_modules`, it is important to have a mechanism for preventing the default Yarn workspace behavior. By adding workspace pattern strings here, Yarn will resume non-workspace behavior for any package which matches the defined patterns.
  [Read more](https://classic.yarnpkg.com/blog/2018/02/15/nohoist/)
  */
  nohoist?: WorkspacePattern[];
}

/**
A workspace pattern points to a directory or group of directories which contain packages that should be included in the workspace installation process.
The patterns are handled with [minimatch](https://github.com/isaacs/minimatch).
@example
`docs` → Include the docs directory and install its dependencies.
`packages/*` → Include all nested directories within the packages directory, like `packages/cli` and `packages/core`.
*/
export type WorkspacePattern = string;

export type YarnConfiguration = {
  /**
  Used to configure [Yarn workspaces](https://classic.yarnpkg.com/docs/workspaces/).
  Workspaces allow you to manage multiple packages within the same repository in such a way that you only need to run `yarn install` once to install all of them in a single pass.
  Please note that the top-level `private` property of `package.json` **must** be set to `true` in order to use workspaces.
  */
  workspaces?: WorkspacePattern[] | WorkspaceConfig;

  /**
  If your package only allows one version of a given dependency, and you’d like to enforce the same behavior as `yarn install --flat` on the command-line, set this to `true`.
  Note that if your `package.json` contains `"flat": true` and other packages depend on yours (e.g. you are building a library rather than an app), those other packages will also need `"flat": true` in their `package.json` or be installed with `yarn install --flat` on the command-line.
  */
  flat?: boolean;

  /**
  Selective version resolutions. Allows the definition of custom package versions inside dependencies without manual edits in the `yarn.lock` file.
  */
  resolutions?: Dependencies;
}

export type JSPMConfiguration = {
  /**
  JSPM configuration.
  */
  jspm?: PackageJSON;
}

export type ESLintConfiguration = {
	/**
	 * Configuration settings for eslint.
	 */
	eslintConfig?: ESLintConfig;
}

export type LicenseID = LiteralUnion<"MIT" | "ISC", string>;

/**
Type for [npm's `package.json` file](https://docs.npmjs.com/creating-a-package-json-file). Also includes types for fields used by other popular projects, like TypeScript and Yarn.
*/
export type PackageJSON = JSONObject & {
	/**
	The name of the package.
	*/
	name?: string;

	/**
	Package version, parseable by [`node-semver`](https://github.com/npm/node-semver).
	*/
	version?: string;

	/**
	Package description, listed in `npm search`.
	*/
	description?: string;

	/**
	Keywords associated with package, listed in `npm search`.
	*/
	keywords?: string[];

	/**
	The URL to the package's homepage.
	*/
	homepage?: LiteralUnion<".", string>;

	/**
	The URL to the package's issue tracker and/or the email address to which issues should be reported.
	*/
	bugs?: BugsLocation;

	/**
	The license for the package.
	*/
	license?: LicenseID;

	/**
	The licenses for the package.
	*/
	licenses?: Array<{
		type?: LicenseID;
		url?: string;
	}>;

	author?: Person;

	/**
	A list of people who contributed to the package.
	*/
	contributors?: Person[];

	/**
	A list of people who maintain the package.
	*/
	maintainers?: Person[];

	/**
	The files included in the package.
	*/
	files?: string[];

	/**
	The module ID that is the primary entry point to the program.
	*/
	main?: string;

	/**
	The executable files that should be installed into the `PATH`.
	*/
	bin?: string | {
		[binary: string]: string;
	};

	/**
	Filenames to put in place for the `man` program to find.
	*/
	man?: string | string[];

	/**
	Indicates the structure of the package.
	*/
	directories?: DirectoryLocations;

	/**
	Location for the code repository.
	*/
	repository?: Repository;

	/**
	Script commands that are run at various times in the lifecycle of the package. The key is the lifecycle event, and the value is the command to run at that point.
	*/
	scripts?: Scripts;

	/**
	Is used to set configuration parameters used in package scripts that persist across upgrades.
	*/
	config?: JSONObject;

	/**
	The dependencies of the package.
	*/
	dependencies?: Dependencies;

	/**
	Additional tooling dependencies that are not required for the package to work. Usually test, build, or documentation tooling.
	*/
	devDependencies?: Dependencies;

	/**
	Dependencies that are skipped if they fail to install.
	*/
	optionalDependencies?: Dependencies;

	/**
	Dependencies that will usually be required by the package user directly or via another dependency.
	*/
	peerDependencies?: Dependencies;

	/**
	Indicate peer dependencies that are optional.
	*/
	peerDependenciesMeta?: {
		[packageName: string]: {
			optional: true;
		};
	};

	/**
	Package names that are bundled when the package is published.
	*/
	bundledDependencies?: string[];

	/**
	Alias of `bundledDependencies`.
	*/
	bundleDependencies?: string[];

	/**
	Engines that this package runs on.
	*/
	engines?: {
		[EngineName in "npm" | "node" | string]: string;
	};

	/**
	@deprecated
	*/
	engineStrict?: boolean;

	/**
	Operating systems the module runs on.
	*/
	os?: Array<LiteralUnion<
		| "aix"
		| "darwin"
		| "freebsd"
		| "linux"
		| "openbsd"
		| "sunos"
		| "win32"
		| "!aix"
		| "!darwin"
		| "!freebsd"
		| "!linux"
		| "!openbsd"
		| "!sunos"
		| "!win32",
		string
	>>;

	/**
	CPU architectures the module runs on.
	*/
	cpu?: Array<LiteralUnion<
		| "arm"
		| "arm64"
		| "ia32"
		| "mips"
		| "mipsel"
		| "ppc"
		| "ppc64"
		| "s390"
		| "s390x"
		| "x32"
		| "x64"
		| "!arm"
		| "!arm64"
		| "!ia32"
		| "!mips"
		| "!mipsel"
		| "!ppc"
		| "!ppc64"
		| "!s390"
		| "!s390x"
		| "!x32"
		| "!x64",
		string
	>>;

	/**
	If set to `true`, a warning will be shown if package is installed locally. Useful if the package is primarily a command-line application that should be installed globally.
	@deprecated
	*/
	preferGlobal?: boolean;

	/**
	If set to `true`, then npm will refuse to publish it.
	*/
	private?: boolean;

	/**
	A set of config values that will be used at publish-time. It's especially handy to set the tag, registry or access, to ensure that a given package is not tagged with 'latest', published to the global public registry or that a scoped module is private by default.
	*/
	publishConfig?: JSONObject;

	/**
	Describes and notifies consumers of a package's monetary support information.
	[Read more.](https://github.com/npm/rfcs/blob/latest/accepted/0017-add-funding-support.md)
	*/
	funding?: string | {
		/**
		The type of funding.
		*/
		type?: LiteralUnion<
			| "github"
			| "opencollective"
			| "patreon"
			| "individual"
			| "foundation"
			| "corporation",
			string
		>;
		/**
		The URL to the funding page.
		*/
		url: string;
	};
} &
NonStandardEntryPoints &
TypeScriptConfiguration &
YarnConfiguration &
JSPMConfiguration &
JSONObject &
ESLintConfiguration;
